europython2020/talk.rst

```diff
@@ -0,0 +1,374 @@
+
+----
+
+``gast``
+========
+
+https://github.com/serge-sans-paille/gast
+
+    ``['gast]`` *f* - good time girl
+
+
+Tool started in 2016 and presented at PyConFR 2016
+
+Originally meant to ease the transition of Pythran to Python3
+
+Core component of AST manipulation:
+
+> Downloads last day: 155,526
+
+*(don't trust numbers)*
+
+
+----
+
+A Simple Expression Across the Ages
+===================================
+
+.. code-block:: console
+
+    > python -c "import ast; print(ast.dump(ast.parse('a[1, ...]')))"
+    ...
+
+----
+
+In Python2
+==========
+
+.. code-block::
+
+    Module(body=[
+        Expr(value=
+                Subscript(value=Name(id='a', ctx=Load()),
+                          slice=
+                            ExtSlice(dims=
+                                    [Index(value=Num(n=1)), Ellipsis()]),
+                          ctx=Load()))
+    ])
+
+----
+
+In Python3.6
+============
+
+.. code-block::
+
+    Module(body=[
+        Expr(value=
+                Subscript(value=Name(id='a', ctx=Load()),
+                          slice=
+                            Index(value=Tuple(elts=[Num(n=1), Ellipsis()],
+                                  ctx=Load())),
+                          ctx=Load()))
+    ])
+
+----
+
+In Python3.9
+============
+
+.. code-block::
+
+    Module(body=[
+        Expr(value=
+                Subscript(value=Name(id='a', ctx=Load()),
+                          slice=
+                            Tuple(elts=[Constant(value=1), Constant(value=Ellipsis)],
+                                  ctx=Load()),
+                          ctx=Load()))
+    ], type_ignores=[])
+
+----
+
+With ``gast``
+=============
+
+
+Whatever the Python version
+
+.. code-block:: console
+
+    > python -c "import gast as ast; \
+                 print(ast.dump(ast.parse('a[1, ...]')))"
+    Module(body=[Expr(value=Subscript(value=Name(id='a', ctx=Load(),
+    annotation=None, type_comment=None), slice=Tuple(elts=[Constant(value=1,
+    kind=None), Constant(value=Ellipsis, kind=None)], ctx=Load()),
+    ctx=Load()))], type_ignores=[])
+
+----
+
+Tradeoffs
+=========
+
+1. Slightly more verbose than Python 3.9 because of Python2 compatibility
+2. Extra translation step, slight performance impact when parsing the world
+
+----
+
+``beniget``
+===========
+
+https://github.com/serge-sans-paille/beniget
+
+    ``['beniget]`` *adj.* - blessed
+
+Compute use-def chains for Python
+
+Foundation of several ``Pythran`` analyses
+
+----
+
+About Use-Def Chains
+====================
+
+From https://en.wikipedia.org/wiki/Use-define_chain
+
+    A Use-Definition Chain (UD Chain) is a data structure that consists of a use, U,
+    of a variable, and all the definitions, D, of that variable that can reach that
+    use without any other intervening definitions
+
+----
+
+Typical Usage
+=============
+
+A def *without* a use means a useless def:
+
+- Detect unused imports
```

Comment:
is useless unless you have expected side-effects in your imports (in `__init__.py`)… but I would not recommend relying on this too much ;)

Reply:
Correct! I'll mention int during the talk :-)

Reply:
Sometimes you try to import a module just to check it's there too